castxml 0.1+git20160202-1 (new formula)

iMichka opened the pull request:
CastXML is the successor of GCC-XML.

CastXML is a tool based on LLVM/Clang that generates the same XML output format as GCC-XML.
It is used by multiple projects like pygccxml and ITK (and all projects using py++ to build python bindings)

The version number follows the one from the debian package (https://packages.debian.org/sid/castxml)

iMichka commented:
The Travis build failed to a timeout because this formula needs ```llvm --with-clang```, so it triggered a llvm recompilation. The homebrew bots are green :)

DomT4 commented on Library/Formula/castxml.rb:
```diff
@@ -0,0 +1,22 @@
+class Castxml < Formula
+  desc "C-family Abstract Syntax Tree XML Output"
+  homepage "https://github.com/CastXML/CastXML"
+  url "https://github.com/CastXML/CastXML/", :revision => "1489405912947356574477a3c10fec69e78cb6dc", :using => :git
```
We'll need an actual stable upstream tag here.

DomT4 replied:
And use `.git` on the end of GitHub URLs and `head`s, rather than passing `:using => :git`, Thanks!

DomT4 commented on Library/Formula/castxml.rb:
```diff
@@ -0,0 +1,22 @@
+  head "https://github.com/CastXML/CastXML/"
+
+  depends_on "cmake" => :build
+  depends_on "llvm" => ["with-clang"]
```
Don't need the `[]` if it's just one conditional.

iMichka commented:
Hi

I adressed the code styling comments.

For the stable upstream tag, this may be complicated. I am not saying there will never be one; but I do not know when it will come.

I already asked upstream for a version tag. The debian packagers just picked a tag (0.1+git20160202-1), so I took the same to be consistent. I hoped that taking the debian version number would be enough
to get this accepted.

I am just the "package maintainer" here so I have not much influence on this. I understand homebrew's point of view on that matter and also personnaly prefer a version tag (my own OSS projects get tags).

I am aware the homebrew-head tap has been put to read-only.
I wrote a comment about this: https://github.com/Homebrew/homebrew-head-only/issues/186#issuecomment-182967897

There are projects out there were each commit in the master branch is a "tag", and there will be no version tag. Each commit is then just considered as a "stable" version.

I am not sure how much the homebrew mainterainers would be willing to negociate on non tagged versions. If there were already a lot of discussion about this point before, and the decision is definitive, I do not want you to loose your time on this.

I can still put castxml in a private tap, so there is no harm if you reject the formula. I am not sure if we should discuss this here or somewhere else, as this is not a castxml specific problem.

mikemcquaid commented:
> I am not sure how much the homebrew mainterainers would be willing to negociate on non tagged versions. If there were already a lot of discussion about this point before, and the decision is definitive, I do not want you to loose your time on this.
> 
> I can still put castxml in a private tap, so there is no harm if you reject the formula. I am not sure if we should discuss this here or somewhere else, as this is not a castxml specific problem.

I think a private tap would be perfect. Our moving away from head-only is because it's a poor user experience and it's really hard for us to debug and fix issues. This is easier for Debian because they vendor the source and use it to build binaries they distribute so nothing is moved from under them.

In this particular case I'm definitely open to other maintainer views though; if upstream considers every commit to be stable and Debian packages it as-is then I can see some flexibility being useful then.

DomT4 commented:
I'd be happier if we used a Debian tarball like https://mirrors.ocf.berkeley.edu/debian/pool/main/c/castxml/castxml_0.1+git20160202.orig.tar.xz rather than a git checkout for stability. I could live with that.

iMichka commented:
For this particular formula, there will hopefully not be too much updates. It would be crazy to update after each commit. The debian packagers have only had 5 versions since June 2015. The project is quite mature now and things have settled down. The only thing that can break it is an incompatible llvm update, which also does not happen to often.

I am OK for using the Debian tarball. Please tell me if I should update the formula accordingly. I'll wait a little bit on other maintainers feedback before making the change.

mikemcquaid commented:
:+1: to Debian tarball.

iMichka commented:
The 10.9 Travis build succeeded. 10.10 and 10.11 timed out (There is around 5 minutes difference between these builds, too bad).

The yosemite brew bot failed because it complains about the version number. The two other bots seem happy though.

DomT4 commented on Library/Formula/castxml.rb:
```diff
@@ -0,0 +1,24 @@
+class Castxml < Formula
+  desc "C-family Abstract Syntax Tree XML Output"
+  homepage "https://github.com/CastXML/CastXML"
+  url "http://http.debian.net/debian/pool/main/c/castxml/castxml_0.1+git20160202.orig.tar.xz"
```
Please don't use a plaintext tarball for the url. Debian has enough secure mirrors to use one of 'em.

DomT4 commented on Library/Formula/castxml.rb:
```diff
@@ -0,0 +1,24 @@
+  end
+
+  test do
+    system "castxml"
```
Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

iMichka commented:
I changed the urls to use the same https mirrors as other formulas. At first I had trouble to find https mirrors, as the debian mirror page has only a list of ftp mirrors (https://www.debian.org/mirror/list) ... and the debian package pages all link to http mirrors ...

I also changed the test to use castxml to parse a simple c++ file. There is an extensive test suite but we probably don't want to run everything here, so this minimal (fast) test should be hopefully sufficient.

dunn commented:
Almost there!
```
==> audit problems
castxml:
 * `version` (line 7) should be put before `checksum` (line 6)
 * Use `system "#{bin}/castxml", "-c", "-x", "c++", "--castxml-cc-gnu", "clang++", "--castxml-gccxml", "-o", "test.xml", "test.cpp"` instead of `system "#{bin}/castxml -c -x c++ --castxml-cc-gnu clang++ --castxml-gccxml -o test.xml test.cpp"`

iMichka commented:
My bad, I forgot the ```--strict``` flag when using ```brew audit```. Maybe strict should just be the default for everybody.

dunn commented:
Thanks! Merged in d212ffd39ba6d1fe9c01cca6589ce651b00fb894.

iMichka commented:
Thanks. To you all. For the support and for your patience. :)
